src/rsa/blinding.rs

```diff
@@ -0,0 +1,585 @@
+        {
+            let mut result = [0, 0];
+            assert!(range.sample_into_limbs(&mut result, &random_ff).is_err());
+        }
+
+        // Test that a generated value exactly equal `max_exclusive` is
+        // rejected and that `generate` gives up after a while of only getting
+        // that value from the PRNG.
+        let max_exclusive_bytes = limbs_as_bytes(&max_exclusive);
+        {
+            let rng = rand::test_util::FixedSliceRandom {
+                bytes: &max_exclusive_bytes
+            };
+            let mut result = [0, 0];
+            assert!(range.sample_into_limbs(&mut result, &rng).is_err());
+        }
+
+        let max_exclusive_minus_1 = [max_exclusive[0] - 1, max_exclusive[1]];
+
+        // Test that a generated value exactly equal to `mex_exclusive - 1` is
+        // accepted.
+        let max_exclusive_minus_1_bytes =
+            limbs_as_bytes(&max_exclusive_minus_1);
+        {
+            let rng = rand::test_util::FixedSliceRandom {
+                bytes: max_exclusive_minus_1_bytes
+            };
+            let mut result = [0, 0];
+            range.sample_into_limbs(&mut result, &rng).unwrap();
+            assert_eq!(&max_exclusive_minus_1, &result);
+        }
+
+        // Test recovery from initial RNG failure.
+        {
+            let bytes = [
+                &max_exclusive_bytes[..],
+                &[0u8; 2 * LIMB_BYTES],
+                &max_exclusive_minus_1_bytes[..],
+            ];
+            let rng = rand::test_util::FixedSliceSequenceRandom {
+                bytes: &bytes,
+                current: core::cell::UnsafeCell::new(0),
+            };
+            let mut result = [0, 0];
+            range.sample_into_limbs(&mut result, &rng).unwrap();
+            assert_eq!(&max_exclusive_minus_1, &result);
+        }
+    }
+}
+
+#[cfg(feature = "internal_benches")]
+mod bench {
+    use {bench, rand};
+    use limb::*;
+    use super::{RSA_KEY_MAX_LIMBS, Range, SamplingParams, with_extended_buffers};
+
+    #[bench]
+    fn bench_sample_into_limbs_simple(b: &mut bench::Bencher) {
+        let mut max: [Limb; RSA_KEY_MAX_LIMBS] = [0; RSA_KEY_MAX_LIMBS];
+        max[RSA_KEY_MAX_LIMBS - 1] = 1 << (LIMB_BITS - 1);
+
+        let mut dest: [Limb; RSA_KEY_MAX_LIMBS] = [0; RSA_KEY_MAX_LIMBS];
+        let rng = rand::SystemRandom::new();
+
+        let params = SamplingParams {
+            most_sig_limb_mask: Limb::max_value(),
+            reduce_when_over_bound: false,
+            extend_limbs_by_one: false,
+        };
+        let range = Range {
+            max_exclusive: &max,
+            sampling_params: &params,
+        };
+
+        b.iter(|| {
+            range.sample_into_limbs(&mut dest, &rng)
+        });
+    }
+
+    #[bench]
+    fn bench_sample_into_limbs_reduce_no_copy(b: &mut bench::Bencher) {
+        let mut max: [Limb; RSA_KEY_MAX_LIMBS] = [0; RSA_KEY_MAX_LIMBS];
+        max[RSA_KEY_MAX_LIMBS - 1] = 1 << (LIMB_BITS - 2);
```

Comment:
Please add a comment and an assertion such as `assert_eq!(max[RSA_KEY_MAX_LIMBS - 1] >> LIMB_BITS - 3, 0b100)` to indicate the expected form of `max`.